### ISSUE TITLE
Faster double-to-string operation when writing out JSON

### ISSUE THREAD

**Issue by michaeleisel** [simdjson/simdjson#1692]
I'm not sure how high of a priority the speed of writing out JSON is, but if it is, I have a prototype that runs in about 40% the time of the fastest library I know of, which is https://github.com/ulfjack/ryu (to be fair ryu, tries to cover a larger set of use cases than we need). In many ways it's just an inverse of what we do for string to double, and uses the same multiplication trick.

**Comment by lemire** [comment]
@michaeleisel Yes. This is a high priority. I started work on a reversed operation, and @nigeltao may also been doing some work (he announced it) but if you got a bit further to a benchmarkable prototype, let us move this forward. Your prototypes have a history of being fruitful.

**Comment by lemire** [comment]
The competitors are ryu, Grisu (which we use in simdjson, we use Grisu 2), Schubfach, Dragonbox...

(see https://github.com/jk-jeon/fp)


But we don't actually care that much about speed per se... of course, any reverse algo. will be fast and efficient, but at least speaking for myself, the motivation is to do both directions with the same code, which helps reduce bloat. This would be the benefit we would have for simdjson.

**Comment by lemire** [comment]
Should also be of interest to  @jpmag

**Comment by lemire** [comment]
That is, for version 1.0 of simdjson, if we could trim out a large chunk of code, it would be a big plus, even if there is no speed benefit at all.

**Comment by lemire** [comment]
Tentatively marked for simdjson 1.0.

**Comment by michaeleisel** [comment]
I'm not sure how much code sharing we can do exactly, but here's the meat of it: https://gist.github.com/michaeleisel/f7b6ece0587bf982895d1eb0bf2b2aa8

There are 3 cases IIRC:
- you can go straight from double to int with a shift, if it fits in a 64-bit unsigned int
- the multiplication trick
- slow fallback

this code is focused on the middle case. IIRC its only outstanding issue for that case is printing out 17 digits, even when a smaller number of digits is sufficient to convert back to a double

**Comment by lemire** [comment]
@michaeleisel Great. I'm otherwise preoccupied right at the moment, but I will start from your code and keep you posted.

**Comment by lemire** [comment]
I think I will be able to build on this later this week.

**Comment by michaeleisel** [comment]
Here's the full project: https://github.com/michaeleisel/floats (can you build with Xcode?)

Make sure to add ryu as a sibling directory

**Comment by lemire** [comment]
Thanks for the pointer. I think that your gist was already great. 

I’ll get started soon.

**Comment by lemire** [comment]
I am marking this for 2.0, removing the 1.0.

**Comment by lemire** [comment]
In my comment above, when I was referring to "reversed operation" and "trimming out a large chunk of code", what I had in mind was a tightly integrated serializer that would reuse the data and code from the deserializer. I see that it is not what your prototype does. It seems that you have built fast code path for the deserializer (from_chars) but it is not directly related to our existing code, it is more of a case where it might compete with our from_chars. So you might have a fast routine for common cases...

**Comment by michaeleisel** [comment]
Agreed
